_posts/2022-04-26-type-based-optimizations-in-the-jit.md

```diff
@@ -0,0 +1,814 @@
+
+We aim to improve the type analysis and optimizations in OTP 26 and
+generate better code for this example. We are also considering adding
+a new guard BIF in OTP 26 for testing that a term is an integer in a
+given range.
+
+Meanwhile, while we wait for OTP 26, there is a way in
+OTP 25 to write an equivalent guard that will result in
+much more efficient code **and** establish known ranges for `X` and
+`Y`:
+
+```assembly
+add5(X, Y) when X =:= X band 16#3FF,
+                Y =:= Y band 16#3FF ->
+    X + Y.
+```
+
+We are showing this way of writing guard for illustrative purposes
+only; we don't recommend rewriting your guards in this way.
+
+The `band` operator fails if not both of its operands are integers, so
+no `is_integer/1` test is needed. The `=:=` comparison will return
+`false` if the corresponding variable is outside the range `0` through
+`16#3FF`.
+
+That will result in the following BEAM code, where the compiler now
+has been able to figure out the possible ranges for the operands of
+the `+` operator:
+
+```
+    {gc_bif,'band',{f,21},2,[{x,0},{integer,1023}],{x,2}}.
+    {test,is_eq_exact,
+          {f,21},
+          [{tr,{x,0},{t_integer,any}},{tr,{x,2},{t_integer,{0,1023}}}]}.
+    {gc_bif,'band',{f,21},2,[{x,1},{integer,1023}],{x,2}}.
+    {test,is_eq_exact,
+          {f,21},
+          [{tr,{x,1},{t_integer,any}},{tr,{x,2},{t_integer,{0,1023}}}]}.
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,{0,1023}}},{tr,{x,1},{t_integer,{0,1023}}}],
+            {x,0}}.
+    return.
+```
+
+Also, the 4 instructions that precede the `+` instructions are now
+relatively efficient.
+
+The `band` instruction needs to test the operands and be prepared to handle
+integers that don't fit in 60 bits:
+
+```assembly
+# i_band_ssjd
+    mov rsi, qword ptr [rbx]
+    mov eax, 16383
+# is the operand small?
+    mov edi, esi
+    and edi, 15
+    cmp edi, 15
+    short jne L97
+    and rax, rsi
+    short jmp L98
+L97:
+    call 4456532680
+    short je label_25
+L98:
+    mov qword ptr [rbx+16], rax
+```
+
+The `is_eq_exact` instruction benefits from type information derived from
+executing the `band` instruction. Since the right-hand side operand is known
+to be a small integer that fits in a machine word, a simple comparison is
+sufficient with no need for fallback code to handle other Erlang terms:
+
+```assembly
+# is_eq_exact_fss
+# simplified check since one argument is an immediate
+    mov rdi, qword ptr [rbx+16]
+    cmp qword ptr [rbx], rdi
+    short jne label_25
+```
+
+The JIT generates the following code for the `+` operator:
+
+```assembly
+# i_plus_ssjd
+# add without overflow check
+    mov rax, qword ptr [rbx]
+    mov rsi, qword ptr [rbx+8]
+    and rax, -16
+    add rax, rsi
+    mov qword ptr [rbx], rax
+```
+
+### Simplifications for `base64`
+
+As far as we know, `base64` is the module in OTP that has benefitted
```

Comment:
benefitted -> benefited

Reply:
"benefitted" is correct, but "benefited" is more common in US English, so I'll change.

Reply:
Yes, after looking this up myself (which I should've before the comment) both are correct, but with some bias of US English vs UK English.